source/MRMesh/MRFillContours2D.cpp

```diff
@@ -0,0 +1,215 @@
+    Vector3f centerPoint;
+    size_t pointsNumber = 0;
+    for ( const auto& contour : contours3f )
+    {
+        const auto max = contour.size() - 1;
+        pointsNumber += max;
+        for ( int i = 0; i < max; ++i )
+            centerPoint += contour[i];
+    }
+    centerPoint = centerPoint / float( pointsNumber );
+
+    // calculate plane normal
+    Vector3f planeNormal;
+    for ( const auto& contour : contours3f )
+    {
+        const auto max = contour.size() - 1;
+        for ( int i = 0; i < max; ++i )
+        {
+            planeNormal += cross( ( contour[i] - centerPoint ), ( contour[i + 1] - centerPoint ) );
+        }
+    }
+    planeNormal = planeNormal.normalized();
+
+    // find transformation from world to plane space and back
+    const auto planeXf = AffineXf3f( Matrix3f::rotation( Vector3f::plusZ(), planeNormal ), centerPoint );
+    const auto planeXfInv = planeXf.inverse();
+
+    // make contours2D (on plane) from contours3D (in world)
+    Contours2f contours2f;
+    float maxZ = 0.f; // FOR DEBUG
+    for ( const auto& contour3f : contours3f )
+    {
+        Contour2f contour;
+        for ( const auto& point : contour3f )
+        {
+            const auto localPoint = planeXfInv( point );
+            contour.push_back( Vector2f( localPoint.x, localPoint.y ) );
+            if ( std::fabs( localPoint.z ) > maxZ ) maxZ = std::fabs( localPoint.z ); // DEBUG
+        }
+        contours2f.push_back( contour );
+    }
+    spdlog::info( "DEBUG maxZ = {}", maxZ ); // DEBUG
+
+    // make patch surface
+    auto fillResult = PlanarTriangulation::triangulateContours( contours2f, true );
+    if ( !fillResult )
+    {
+        spdlog::warn( "Cant triangulate contours" );
+        return false;
+    }
+    Mesh& additionalMesh = *fillResult;
+    auto holes = additionalMesh.topology.findHoleRepresentiveEdges();
+    auto holesForSearch = holes;
+
+    // transform patch surface from plane to world space
+    auto& addMeshPoints = additionalMesh.points;
+    for ( auto& point : addMeshPoints )
+        point = planeXf( point );
+    
+    
+    // find 
+    std::vector<Vector2i> mapToSourceMesh( paths.size(), Vector2i( -1, -1 ) );
+    for ( int i = 0; i < paths.size(); ++i )
+    {
+        auto& path = paths[i];
+        for ( int j = 0; j < path.size(); ++j )
+        {
+            auto sourceOrg = mesh.orgPnt( path[j] );
+            auto sourceDest = mesh.destPnt( path[j] );
+            bool found = false;
+            for ( int k = int( holesForSearch.size() ) - 1; k >= 0; --k )
+            {
+                auto addOrg = additionalMesh.orgPnt( holesForSearch[k] );
+                auto addDest = additionalMesh.destPnt( holesForSearch[k] );
+                found = ( ( sourceOrg - addOrg ).lengthSq() < 1.e-6f && ( sourceDest - addDest ).lengthSq() < 1.e-6f ) ||
+                    ( ( sourceOrg - addDest ).lengthSq() < 1.e-6f && ( sourceDest - addOrg ).lengthSq() < 1.e-6f );
```

Comment:
better use std::numeric_limits<float>::epsilon() (also, not this map identical: (1,1), (2,2) ... ?)

Reply:
reworked